scroll-animations-1/Overview.bs

```diff
@@ -353,18 +354,105 @@ spec: cssom-view-1; type: dfn;
 
 	Values are as defined for ''scroll()''.
 
+	The [=used value=] of 'scroll-timeline-axis'
+	is the axis of the [=current attachment=],
+	or the initial value if there is no [=current attachment=].
+
+### Attachment of a Scroll Progress Timeline: the 'scroll-timeline-attachment' property ### {#scroll-timeline-attachment}
+
+	<pre class='propdef'>
+	Name: scroll-timeline-attachment
+	Value: [ local | defer | ancestor ]#
+	Initial: local
+	Applies to: all elements
+	Inherited: no
+	Computed value: a list of the keywords specified
+	Animation type: not animatable
+	</pre>
+
+	Specifies attachment behavior for each [=scroll progress timeline=]
+	associated with this element.
+
+	A scroll timeline has a set of <dfn>attachments</dfn>,
+	that determine which [=scroll container=] and {{ScrollTimeline/axis}} to use
+	when calculating the [progress of the timeline](#scroll-timeline-progress).
```

Comment:
Right, it seemed logical to me that `scroll-timeline-*` would actually define and create scroll timelines (regardless of their nature), and then `scroll-timeline-attachment` would describe how these timelines attach together (it's what it sounds like, isn't it?).

Let me give you some minimal background about this solution, _not_ to "prove" that "scope the name" isn't a viable way to spec (maybe it is), but to help you write the best possible spec. :-)

Suppose that we had elements A, B and C that appear in that order in the tree.

```
  A {
    scroll-timeline: mytimeline defer;
  }
  A B {
    animation: anim auto;
    animation-timeline: mytimeline;
  }
  A B + C {
    scroll-timeline: mytimeline ancestor;
  }
```

 - First we resolve the style for A. In "scope the name", this yields no timeline.
 - Then we resolve the style for B. Doing so creates CSS animations that have no timeline, since we can't find "mytimeline".
 - Then we resolve the style for C. This creates "mytimeline", which is now reachable via A.

But animations from B already decided that they have no timeline, and have no memory or knowledge that they should _now_ be connected to some timeline after all. At least, they don't without additional spec concepts, and explicitly deferred timelines was an attempt at specifying such a concept!

My support for this feature was predicated on attachment hackery being an [aspect of the timeline (comment)](https://github.com/w3c/csswg-drafts/issues/7759#issuecomment-1413582984), and I still think this is the only realistically possible way to do it. In particular, animations should be allowed to remain agnostic of said hackery without having its timeline replaced in some awkward "fixup" step.

Reply:
> Also, multiple attachments should not be a fundamental data structure in CSS, they're an error condition. You can implement error handling data structures however you want internally, but let's not write the spec as if they're fundamentally how this works.

Done, FWIW.

Reply:
I'm stating the obvious maybe, but having deferred names affect timeline lookups (i.e. affect what animation-timeline sees) creates a circular dependency within the same frame.

```
<A>
  <B>
  </B>
</A>
```

```
@keyframes anim {
  from { --x:0; }
  to { --x:0; }
}

A {
  --x:1;
  scroll-timeline: timeline defer;
  animation: anim auto;
  animation-timeline: timeline;
}

@container style(--x:1) {
  B {
    scroll-timeline: timeline ancestor;
  }
}
```

If the computed value of `--x` on `A` is `1`, then timeline on `B` should have been reachable via the deferred name on `A`, and `--x` on `A` must be `0`.

(This is a problem with or without style queries, but style queries makes it more obvious).

I said earlier that maybe there is a way to make this work as "scopes as a name", but I'm increasingly convinced that it's not.

Reply:
I'm not sure how changing how we describe it affects this case? In your text, the timeline is inactive until it's bound; doesn't that mean the keyframes aren't applied?

Reply:
In your PR, deferred names _don't_ affect timeline lookups, so it's fine. I didn't know you were just going to change how we describe it. :-)

> In your text, the timeline is inactive until it's bound; doesn't that mean the keyframes aren't applied?

In the first frame, the timeline is inactive because it doesn't have a state yet. This state is updated at a specific point in the frame (before the style update), and we're already beyond that point for the _current_ frame. When we resolve the style for B in this frame, timelines attach to each other, but it has no immediate affect on any timeline state. The timeline on A is still inactive.

In the _next_ frame, the state of all scroll timelines are updated. In this moment (only in this moment), the attachments matter, because we use the attached timeline to produce the state. In the subsequent style resolution, the animation applies this time, and we get `--x:0`. This removes the timeline on B, leaving the deferred timeline with no attachments. This (again) does not affect the state of the timeline this frame. The computed value of `--x` is still `0`.

In the next frame after that, the timeline once again becomes inactive, because it has no attachments at the time of the state update. So we actually switch between the two states every frame. But there is no circularity _within the same frame_, as a lookup-affecting version of "scopes a name" would have caused.

I'm not sure if that answers the question ...

(Note that I'm ignoring our handling of [initially stale](https://drafts.csswg.org/scroll-animations-1/#initially-stale) timelines in the above, so what would _really_ happen would require an even longer explanation).

Reply:
> In your PR, deferred names don't affect timeline lookups, so it's fine.

Yeah, because you said that would be bad. :) My first instinct was to have them be invisible, because in CSS we usually ignore things that are invalid (treat them as if they didn't exist)...

> I'm not sure if that answers the question ...

It does. What a terrible scenario.